File: examples/latest/navigate/android/MapFeatures/app/src/main/java/com/here/mapfeatures/MainActivity.java
/*
 * Copyright (C) 2019-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.mapfeatures;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.here.sdk.core.GeoCoordinates;
import com.here.sdk.core.engine.AuthenticationMode;
import com.here.sdk.core.engine.SDKNativeEngine;
import com.here.sdk.core.engine.SDKOptions;
import com.here.sdk.core.errors.InstantiationErrorException;
import com.here.sdk.mapview.MapError;
import com.here.sdk.mapview.MapMeasure;
import com.here.sdk.mapview.MapScene;
import com.here.sdk.mapview.MapScheme;
import com.here.sdk.mapview.MapView;

import java.util.Map;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private PermissionsRequestor permissionsRequestor;
    private MapView mapViewGlobe, mapViewWebMercator;
    private MapFeaturesExample mapFeaturesExample;
    private MapSchemesExample mapSchemesExample;
    private MapScene mapScene;
    private Button webMercatorButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Usually, you need to initialize the HERE SDK only once during the lifetime of an application.
        initializeHERESDK();

        setContentView(R.layout.activity_main);

        // Get a MapView instance from layout
        mapViewGlobe = findViewById(R.id.map_view_globe);
        mapViewGlobe.setVisibility(MapView.VISIBLE);
        mapViewWebMercator = findViewById(R.id.map_view_web_mercator);
        webMercatorButton = findViewById(R.id.web_mercator_button);
        mapViewWebMercator.setVisibility(MapView.GONE);
        mapViewWebMercator.onCreate(savedInstanceState);
        mapViewGlobe.onCreate(savedInstanceState);

        webMercatorButton.setOnClickListener(v -> {
            changeMapProjection();
        });

        handleAndroidPermissions();
    }

    private void initializeHERESDK() {
        // Set your credentials for the HERE SDK.
        String accessKeyID = "YOUR_ACCESS_KEY_ID";
        String accessKeySecret = "YOUR_ACCESS_KEY_SECRET";
        AuthenticationMode authenticationMode = AuthenticationMode.withKeySecret(accessKeyID, accessKeySecret);
        SDKOptions options = new SDKOptions(authenticationMode);
        try {
            Context context = this;
            SDKNativeEngine.makeSharedInstance(context, options);
        } catch (InstantiationErrorException e) {
            throw new RuntimeException("Initialization of HERE SDK failed: " + e.error.name());
        }
    }

    private MapView getCurrentVisibleMapView() {
        return mapViewGlobe.getVisibility() == View.VISIBLE ? mapViewGlobe : mapViewWebMercator;
    }

    private void changeMapProjection() {
        Map<String, String> enabledFeatures = mapFeaturesExample.getEnabledFeatures();
        boolean isGlobeVisible = mapViewGlobe.getVisibility() == View.VISIBLE;

        MapView mapViewToHide = isGlobeVisible ? mapViewGlobe : mapViewWebMercator;
        MapView mapViewToShow = isGlobeVisible ? mapViewWebMercator : mapViewGlobe;

        mapViewToHide.setVisibility(View.GONE);
        mapViewToShow.setVisibility(View.VISIBLE);

        if (mapSchemesExample != null) {
            MapScheme currentScheme = mapSchemesExample.getCurrentMapScheme();
            if (currentScheme == null) {
                // Load a default scheme if none is currently set
                currentScheme = MapScheme.NORMAL_DAY;
            }
            mapSchemesExample.loadSchemeForCurrentView(mapViewToShow, currentScheme);
        }

        if (mapFeaturesExample != null) {
            mapFeaturesExample = new MapFeaturesExample(mapViewToShow.getMapScene());
            mapFeaturesExample.applyEnabledFeaturesForMapScene(mapViewToShow.getMapScene());
        }

        for (Map.Entry<String, String> feature : enabledFeatures.entrySet()) {
            mapFeaturesExample.enableFeature(feature.getKey(), feature.getValue());
        }

        webMercatorButton.setText(isGlobeVisible ? "Switch to Globe" : "Switch to Web Mercator");
    }

    private void handleAndroidPermissions() {
        permissionsRequestor = new PermissionsRequestor(this);
        permissionsRequestor.request(new PermissionsRequestor.ResultListener(){

            @Override
            public void permissionsGranted() {
                loadMapScene();
            }

            @Override
            public void permissionsDenied() {
                Log.e(TAG, "Permissions denied by user.");
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        permissionsRequestor.onRequestPermissionsResult(requestCode, grantResults);
    }

    private void loadMapScene() {
        MapView currentMapView = getCurrentVisibleMapView();
        currentMapView.getMapScene().loadScene(MapScheme.NORMAL_DAY, new MapScene.LoadSceneCallback() {
            @Override
            public void onLoadScene(@Nullable MapError mapError) {
                if (mapError == null) {
                    mapSchemesExample = new MapSchemesExample();
                    mapFeaturesExample = new MapFeaturesExample(currentMapView.getMapScene());
                    mapFeaturesExample.applyEnabledFeaturesForMapScene(currentMapView.getMapScene());

                    double distanceInMeters = 1000 * 20;
                    MapMeasure mapMeasureZoom = new MapMeasure(MapMeasure.Kind.DISTANCE_IN_METERS, distanceInMeters);
                    currentMapView.getCamera().lookAt(new GeoCoordinates(52.51760485151816, 13.380312380535472), mapMeasureZoom);
                } else {
                    Log.d(TAG, "onLoadScene failed: " + mapError);
                }
            }
        });
    }

    @Override
    protected void onPause() {
        mapViewGlobe.onPause();
        super.onPause();
    }

    @Override
    protected void onResume() {
        mapViewGlobe.onResume();
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        mapViewGlobe.onDestroy();
        disposeHERESDK();
        super.onDestroy();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        mapViewGlobe.onSaveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    private void disposeHERESDK() {
        // Free HERE SDK resources before the application shuts down.
        // Usually, this should be called only on application termination.
        // Afterwards, the HERE SDK is no longer usable unless it is initialized again.
        SDKNativeEngine sdkNativeEngine = SDKNativeEngine.getSharedInstance();
        if (sdkNativeEngine != null) {
            sdkNativeEngine.dispose();
            // For safety reasons, we explicitly set the shared instance to null to avoid situations,
            // where a disposed instance is accidentally reused.
            SDKNativeEngine.setSharedInstance(null);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater=getMenuInflater();
        inflater.inflate(R.menu.map_option_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        MapView currentMapView = getCurrentVisibleMapView();
        switch (item.getItemId()){
            // Map Schemes:
            case R.id.hybrid_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.HYBRID_DAY);
                return true;
            case R.id.hybrid_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.HYBRID_NIGHT);
                return true;
            case R.id.lite_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LITE_DAY);
                return true;
            case R.id.lite_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LITE_NIGHT);
                return true;
            case R.id.lite_hybrid_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LITE_HYBRID_NIGHT);
                return true;
            case R.id.lite_hybrid_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LITE_HYBRID_DAY);
                return true;
            case R.id.logistics_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LOGISTICS_DAY);
                return true;
            case R.id.logistics_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LOGISTICS_NIGHT);
                return true;
            case R.id.logistics_hybrid_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LOGISTICS_HYBRID_DAY);
                return true;
            case R.id.logistics_hybrid_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.LOGISTICS_HYBRID_NIGHT);
                return true;
            case R.id.normal_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.NORMAL_DAY);
                return true;
            case R.id.normal_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.NORMAL_NIGHT);
                return true;
            case R.id.road_network_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.ROAD_NETWORK_DAY);
                return true;
            case R.id.road_network_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.ROAD_NETWORK_NIGHT);
                return true;
            case R.id.satellite_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.SATELLITE);
                return true;
            case R.id.topo_day_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.TOPO_DAY);
                return true;
            case R.id.topo_night_menu_item:
                mapSchemesExample.loadSchemeForCurrentView(currentMapView, MapScheme.TOPO_NIGHT);
                return true;

            // Map Features:
            case R.id.clear_menu_item:
                mapFeaturesExample.disableFeatures();
                return true;
            case R.id.building_footprints_menu_item:
                mapFeaturesExample.enableBuildingFootprints();
                return true;
            case R.id.congestion_zone_menu_item:
                mapFeaturesExample.enableCongestionZones();
                return true;
            case R.id.environmental_zones_menu_item:
                mapFeaturesExample.enableEnvironmentalZones();
                return true;
            case R.id.extruded_buildings_menu_item:
                mapFeaturesExample.enableExtrudedBuildings();
                return true;
            case R.id.landmarks_textured_menu_item:
                mapFeaturesExample.enableLandmarksTextured();
                return true;
            case R.id.landmarks_textureless_menu_item:
                mapFeaturesExample.enableLandmarksTextureless();
                return true;
            case R.id.safety_cameras_menu_item:
                mapFeaturesExample.enableSafetyCameras();
                return true;
            case R.id.shadows_menu_item:
                mapFeaturesExample.enableShadows();
                return true;
            case R.id.terrain_hillshade_menu_item:
                mapFeaturesExample.enableTerrainHillShade();
                return true;
            case R.id.terrain_3D_menu_item:
                mapFeaturesExample.enableTerrain3D();
                return true;
            case R.id.traffic_flow_with_freeflow_menu_item:
                mapFeaturesExample.enableTrafficFlowWithFreeFlow();
                return true;
            case R.id.traffic_flow_without_freeflow_menu_item:
                mapFeaturesExample.enableTrafficFlowWithoutFreeFlow();
                return true;
            case R.id.traffic_incidents_menu_item:
                mapFeaturesExample.enableTrafficIncidents();
                return true;
            case R.id.vehicle_restrictions_active_menu_item:
                mapFeaturesExample.enableVehicleRestrictionsActive();
                return true;
            case R.id.vehicle_restrictions_active_inactive_menu_item:
                mapFeaturesExample.enableVehicleRestrictionsActiveAndInactive();
                return true;
            case R.id.vehicle_restrictions_active_inactive_diff_menu_item:
                mapFeaturesExample.enableVehicleRestrictionsActiveAndInactiveDiff();
                return true;
            case R.id.road_exit_labels_menu_item:
                mapFeaturesExample.enableRoadExitLabels();
                return true;
            case R.id.road_exit_labels_numbers_menu_item:
                mapFeaturesExample.enableRoadExitLabelsNumbersOnly();
                return true;
            case R.id.ambient_occlusion_menu_item:
                mapFeaturesExample.enableAmbientOcclusion();
                return true;
            case R.id.contours_menu_item:
                mapFeaturesExample.enableContours();
                return true;
            case R.id.low_speed_zones_menu_item:
                mapFeaturesExample.enableLowSpeedZones();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: examples/latest/navigate/android/Kotlin/MapFeaturesKotlin/app/src/main/java/com/here/mapfeatureskotlin/MainActivity.kt
/*
 * Copyright (C) 2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.mapfeatureskotlin

import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import com.here.mapfeatureskotlin.ui.theme.MapFeaturesTheme
import com.here.mapitemskotlin.DropdownMenu
import com.here.sdk.core.GeoCoordinates
import com.here.sdk.core.engine.AuthenticationMode
import com.here.sdk.core.engine.SDKNativeEngine
import com.here.sdk.core.engine.SDKOptions
import com.here.sdk.core.errors.InstantiationErrorException
import com.here.sdk.mapview.MapMeasure
import com.here.sdk.mapview.MapProjection
import com.here.sdk.mapview.MapScheme
import com.here.sdk.mapview.MapView
import com.here.sdk.mapview.MapViewOptions
import com.here.sdk.units.core.utils.EnvironmentLogger

class MainActivity : ComponentActivity() {


    private val environmentLogger = EnvironmentLogger()
    private lateinit var permissionsRequestor: PermissionsRequestor
    private var mapViewGlobe: MapView? = null
    private var mapViewWebMercator: MapView? = null
    private lateinit var mapFeaturesExample: MapFeaturesExample
    private lateinit var mapSchemesExample: MapSchemesExample
    private var isMapViewGlobeVisible by mutableStateOf(false)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Needs to be called before the activity is started.

        // Log application and device details.
        // It expects a string parameter that describes the application source language.
        environmentLogger.logEnvironment("Kotlin")
        permissionsRequestor = PermissionsRequestor(this)

        // Usually, you need to initialize the HERE SDK only once during the lifetime of an application.
        // Before creating a MapView instance please make sure that the HERE SDK is initialized.
        initializeHERESDK()

        enableEdgeToEdge()

        setContent {
            MapFeaturesTheme {
                Scaffold(
                    modifier = Modifier.fillMaxSize(),
                ) { paddingValues ->
                    Box(modifier = Modifier.padding(paddingValues)) {
                        if (isMapViewGlobeVisible) {
                            GlobalMapView(savedInstanceState)
                        } else {
                            MercatorMapView(savedInstanceState)
                        }
                        CustomDropDownMenu()
                        WebMercatorButton()
                    }
                }
            }
        }
    }

    @Composable
    private fun GlobalMapView(savedInstanceState: Bundle?) {
        AndroidView(factory = { context ->
            MapView(context).apply {
                mapViewGlobe = this
                mapViewGlobe?.onCreate(savedInstanceState)

                // Note that for this app handling of permissions is optional as no sensitive permissions
                // are required.
                // Only after permissions have been granted (if any), we load the map view and start the app.
                handleAndroidPermissions()
            }
        })
    }

    @Composable
    private fun MercatorMapView(savedInstanceState: Bundle?) {
        var mapViewOptions = MapViewOptions()
        mapViewOptions.projection = MapProjection.WEB_MERCATOR

        AndroidView(factory = { context ->
            MapView(context, mapViewOptions).apply {
                mapViewWebMercator = this
                mapViewWebMercator?.onCreate(savedInstanceState)

                // Note that for this app handling of permissions is optional as no sensitive permissions
                // are required.
                // Only after permissions have been granted (if any), we load the map view and start the app.
                handleAndroidPermissions()
            }
        })
    }

    @Composable
    fun WebMercatorButton() {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            verticalArrangement = Arrangement.Bottom,
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            CustomButton(
                onClick = {
                    mapFeaturesExample.disableFeatures()
                    isMapViewGlobeVisible = !isMapViewGlobeVisible
                          },
                text = if (isMapViewGlobeVisible) "  Switch to Web Mercator  " else "  Switch to Globe  "
            )
        }
    }

    @Composable
    fun CustomButton(onClick: () -> Unit, text: String) {
        Button(
            onClick = onClick,
            colors = ButtonDefaults.buttonColors(
                containerColor = Color(0xFF005CB9)
            )
        ) {
            Text(text)
        }
    }

    private fun initializeHERESDK() {
        // Set your credentials for the HERE SDK.
        val accessKeyID = "YOUR_ACCESS_KEY_ID"
        val accessKeySecret = "YOUR_ACCESS_KEY_SECRET"
        val authenticationMode = AuthenticationMode.withKeySecret(accessKeyID, accessKeySecret)
        val options = SDKOptions(authenticationMode)
        try {
            val context = this
            SDKNativeEngine.makeSharedInstance(context, options)
        } catch (e: InstantiationErrorException) {
            throw RuntimeException("Initialization of HERE SDK failed: " + e.error.name)
        }
    }

    // Convenience method to check all permissions that have been added to the AndroidManifest.
    private fun handleAndroidPermissions() {
        permissionsRequestor.requestPermissionsFromManifest(
            object : PermissionsRequestor.ResultListener {
                override fun permissionsGranted() {
                    loadMapScene()
                }

                override fun permissionsDenied(deniedPermissions: List<String>) {
                    Log.e(TAG, "Permissions denied by the user.")
                }
            }
        )
    }

    private fun getCurrentVisibleMapView(): MapView? {
        return if (isMapViewGlobeVisible) mapViewGlobe else mapViewWebMercator
    }

    private fun loadMapScene() {
        val mapViewNonNull = (if (isMapViewGlobeVisible) mapViewGlobe else mapViewWebMercator) ?: run {
            Log.e(TAG, "mapView is null. Cannot load map scene.")
            return
        }

        // Load a scene from the HERE SDK to render the map with a map scheme.
        mapViewNonNull.mapScene.loadScene(MapScheme.NORMAL_DAY) { mapError ->
            if (mapError == null) {
                mapSchemesExample = MapSchemesExample()
                mapFeaturesExample = MapFeaturesExample(mapViewNonNull.mapScene)
                mapFeaturesExample.applyEnabledFeaturesForMapScene(mapViewNonNull.mapScene)

                val distanceInMeters = (1000 * 20).toDouble()
                val mapMeasureZoom = MapMeasure(MapMeasure.Kind.DISTANCE_IN_METERS, distanceInMeters)
                mapViewNonNull.camera.lookAt(GeoCoordinates(52.51760485151816, 13.380312380535472), mapMeasureZoom)
            } else {
                Log.d(TAG, "Loading of map failed: mapError: ${mapError.name}")
            }
        }
    }

    @Composable
    fun CustomDropDownMenu() {
        DropdownMenu(
            onMapSchemesClick = { item ->
                when (item) {
                    "Lite Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LITE_NIGHT
                    )
                    "Hybrid Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.HYBRID_DAY
                    )
                    "Hybrid Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.HYBRID_NIGHT
                    )
                    "Lite Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LITE_DAY
                    )
                    "Lite Hybrid Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LITE_HYBRID_DAY
                    )
                    "Lite Hybrid Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LITE_HYBRID_NIGHT
                    )
                    "Logistics Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LOGISTICS_DAY
                    )
                    "Logistics Hybrid Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LOGISTICS_HYBRID_DAY
                    )
                    "Logistics Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LOGISTICS_NIGHT
                    )
                    "Logistics Hybrid Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.LOGISTICS_HYBRID_NIGHT
                    )
                    "Normal Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.NORMAL_DAY
                    )
                    "Normal Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.NORMAL_NIGHT
                    )
                    "Road Network Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.ROAD_NETWORK_DAY
                    )
                    "Road Network Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.ROAD_NETWORK_NIGHT
                    )
                    "Satellite" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.SATELLITE
                    )
                    "Topo Day" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.TOPO_DAY
                    )
                    "Topo Night" -> mapSchemesExample.loadSchemeForCurrentView(
                        getCurrentVisibleMapView(),
                        MapScheme.TOPO_NIGHT
                    )
                }
            },
            onMapFeaturesClick = { item ->
                when (item) {
                    "Clear Map Features" -> mapFeaturesExample.disableFeatures()
                    "Building Footprints" -> mapFeaturesExample.enableBuildingFootprints()
                    "Congestion Zone" -> mapFeaturesExample.enableCongestionZones()
                    "Environmental Zones" -> mapFeaturesExample.enableEnvironmentalZones()
                    "Extruded Buildings" -> mapFeaturesExample.enableExtrudedBuildings()
                    "Landmarks Textured" -> mapFeaturesExample.enableLandmarksTextured()
                    "Landmarks Textureless" -> mapFeaturesExample.enableLandmarksTextureless()
                    "Safety Cameras" -> mapFeaturesExample.enableSafetyCameras()
                    "Shadows" -> {
                        Toast.makeText(
                            this, "Enabled building shadows for non-satellite-based schemes.",
                            Toast.LENGTH_SHORT
                        ).show()
                        mapFeaturesExample.enableShadows()
                    }
                    "Terrain Hillshade" -> mapFeaturesExample.enableTerrainHillShade()
                    "Terrain 3D" -> mapFeaturesExample.enableTerrain3D()
                    "Ambient Occlusion" -> mapFeaturesExample.enableAmbientOcclusion()
                    "Contours" -> mapFeaturesExample.enableContours()
                    "Low Speed Zones" -> mapFeaturesExample.enableLowSpeedZones()
                    "Traffic Flow with Free Flow" -> mapFeaturesExample.enableTrafficFlowWithFreeFlow()
                    "Traffic Flow without Free Flow" -> mapFeaturesExample.enableTrafficFlowWithoutFreeFlow()
                    "Traffic Incidents" -> mapFeaturesExample.enableTrafficIncidents()
                    "Vehicle Restrictions Active" -> mapFeaturesExample.enableVehicleRestrictionsActive()
                    "Vehicle Restrictions Active/Inactive" -> mapFeaturesExample.enableVehicleRestrictionsActiveAndInactive()
                    "Vehicle Restrictions Active/Inactive Diff" -> mapFeaturesExample.enableVehicleRestrictionsActiveAndInactiveDiff()
                    "Road Exit Labels" -> mapFeaturesExample.enableRoadExitLabels()
                    "Road Exit Labels Numbers Only" -> mapFeaturesExample.enableRoadExitLabelsNumbersOnly()
                }
            }
        )
    }

    override fun onPause() {
        mapViewGlobe?.onPause()
        super.onPause()
    }

    override fun onResume() {
        mapViewGlobe?.onResume()
        super.onResume()
    }

    override fun onDestroy() {
        mapViewGlobe?.onDestroy()
        disposeHERESDK()
        super.onDestroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        mapViewGlobe?.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    private fun disposeHERESDK() {
        // Free HERE SDK resources before the application shuts down.
        // Usually, this should be called only on application termination.
        // Afterwards, the HERE SDK is no longer usable unless it is initialized again.
        SDKNativeEngine.getSharedInstance()?.dispose()
        // For safety reasons, we explicitly set the shared instance to null to avoid situations,
        // where a disposed instance is accidentally reused.
        SDKNativeEngine.setSharedInstance(null)
    }

    private companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}
